### PR TITLE
feat: Implement hub interchange validation (#52)

### LINKED ISSUE #52: Backend: Update route validation to allow hub interchanges
## Summary

Modify route validation logic to recognize and allow cross-mode interchanges at hub stations (stations with matching `hub_naptan_code`).

## Related To
- Parent Issue: #47
- Depends On: Issues #50 (schema) and #51 (hub data fetching)

## Problem
Current validation requires direct line connections between consecutive stations. This fails for cross-mode interchanges where users change from one transport mode to another at the same physical location (e.g., Overground → Tube at Seven Sisters).

## Solution
Before checking line connections, verify if consecutive stations share the same `hub_naptan_code`. If they do, allow the transition without requiring a direct line connection.

## Task List

- [ ] Update `validate_route()` with hub interchange logic
- [ ] Add hub code comparison before connection check
- [ ] Add structured logging for hub interchanges
- [ ] Preserve existing branch-aware validation
- [ ] Handle None hub codes gracefully
- [ ] Update error messages if needed
- [ ] Document validation flow in code comments
- [ ] Create comprehensive test cases
- [ ] Ensure pre-commit hooks pass
- [ ] Achieve 100% coverage on new code

### Update validate_route() Method

File: `backend/app/services/tfl_service.py` (around line 1550-1705)

```python
async def validate_route(
    self, segments: list[RouteSegmentRequest]
) -> tuple[bool, str, int | None]:
    # ... existing validation: min segments, duplicates, etc. ...
    
    # Fetch all stations
    stations_map = await self._get_stations_map(
        [seg.station_tfl_id for seg in segments], db
    )
    
    # Check connections between consecutive segments
    for i in range(len(segments) - 1):
        current_segment = segments[i]
        next_segment = segments[i + 1]
        
        from_station = stations_map[current_segment.station_tfl_id]
        to_station = stations_map[next_segment.station_tfl_id]
        
        # NEW: Check for hub interchange (cross-mode transfer)
        if from_station.hub_naptan_code and to_station.hub_naptan_code:
            if from_station.hub_naptan_code == to_station.hub_naptan_code:
                logger.info(
                    "hub_interchange_detected",
                    segment_index=i,
                    from_station=from_station.name,
                    from_id=from_station.tfl_id,
                    to_station=to_station.name,
                    to_id=to_station.tfl_id,
                    hub_code=from_station.hub_naptan_code,
                    from_line=current_segment.line_tfl_id,
                    to_line=next_segment.line_tfl_id
                )
                # Same hub = valid interchange, skip connection check
                continue
        
        # Destination segment - no connection check needed
        if next_segment.line_tfl_id is None:
            continue
        
        # Existing connection validation for non-hub segments
        is_connected = await self._check_connection(
            from_station=from_station,
            to_station=to_station,
            line_tfl_id=next_segment.line_tfl_id,
            db=db,
        )
        
        if not is_connected:
            return (
                False,
                f"No connection found between '{from_station.name}' "
                f"and '{to_station.name}' on line '{next_segment.line_tfl_id}'",
                i
            )
    
    return (True, "Route is valid", None)
```

## Key Logic

1. **Hub Interchange Check First**: Before line connection validation
2. **Both Must Have Hub Codes**: If either is None, skip hub check
3. **Exact Match Required**: `from_station.hub_naptan_code == to_station.hub_naptan_code`
4. **Continue on Match**: Skip connection validation, proceed to next segment
5. **Fall Through**: If no hub match, use existing connection validation

## Testing Requirements

Coverage: 100% on new code

Test cases:
```python
async def test_validate_route_hub_interchange_seven_sisters_rail():
    """Seven Sisters hub interchange using rail ID (910GSEVNSIS)."""
    
async def test_validate_route_hub_interchange_seven_sisters_tube():
    """Seven Sisters hub interchange using tube ID (940GZZLUSVS)."""
    
async def test_validate_route_different_hubs_no_connection_fails():
    """Stations with different hub codes and no connection fail."""
    
async def test_validate_route_no_hub_requires_connection():
    """Stations without hub codes still require valid connections."""
    
async def test_validate_route_one_hub_one_regular_requires_connection():
    """One hub station + one regular station requires connection."""
    
async def test_validate_route_multiple_hub_interchanges():
    """Route with multiple hub interchanges validates correctly."""
    
async def test_validate_route_hub_interchange_logs_correctly():
    """Hub interchange detection logged with all details."""
```

## Test Data Setup

Use Seven Sisters as primary test case:
- `910GSEVNSIS` (rail/overground) → `hub_naptan_code: 'HUBSVS'`
- `940GZZLUSVS` (Victoria line) → `hub_naptan_code: 'HUBSVS'`

Example test route (from issue #47):
```json
{
  "segments": [
    {"station_tfl_id": "910GBHILLPK", "line_tfl_id": "weaver"},
    {"station_tfl_id": "910GSEVNSIS", "line_tfl_id": "victoria"},
    {"station_tfl_id": "940GZZLUPCO", "line_tfl_id": null}
  ]
}
```

Should be equivalent to:
```json
{
  "segments": [
    {"station_tfl_id": "910GBHILLPK", "line_tfl_id": "weaver"},
    {"station_tfl_id": "940GZZLUSVS", "line_tfl_id": "victoria"},
    {"station_tfl_id": "940GZZLUPCO", "line_tfl_id": null}
  ]
}
```

## Files to Modify

- `backend/app/services/tfl_service.py` (validate_route ~line 1550-1705)
- `backend/tests/test_tfl_service.py` (add hub validation tests)

## Validation Steps

1. Ensure migration and hub data fetch complete
2. Rebuild graph: `POST /admin/tfl/build-graph`
3. Test validation via API:
   ```bash
   curl -X POST http://localhost:8000/tfl/validate-route \
     -H "Content-Type: application/json" \
     -d '{"segments": [...]}'
   ```
4. Run tests: `uv run pytest backend/tests/test_tfl_service.py -xvs -k hub`
5. Check logs for hub interchange detection
6. Verify both test cases from issue #47 pass
7. Run pre-commit: `uv run pre-commit run --all-files`

## Dependencies

- **Requires:** #50 (Database schema)
- **Requires:** #51 (Hub data populated)

## Blocks

- Backend testing issue
- Frontend updates

## Coverage Target

- **New Code:** 100%
- **Overall:** Maintain ~95% coverage

## Edge Cases to Handle

- Both stations have hub codes, different hubs → require connection
- One station has hub code, other doesn't → require connection  
- Both stations have hub codes = None → require connection
- Same hub but no valid line transition → allowed (hub logic wins)
- Multiple consecutive hub interchanges → all allowed

## Notes

- Hub logic takes precedence over line connection validation
- Existing branch-aware validation preserved for non-hub segments
- Logging uses structured format for observability
- No OSI (Out-of-Station Interchange) support (YAGNI)

### PR DESCRIPTION
**Pull request by mnbf9rca**
Add hub equivalence logic to route validation, allowing cross-mode interchanges at stations sharing the same hub_naptan_code.

Implementation:
- _get_hub_equivalent_stations(): Query stations by hub code
- _check_any_hub_connection(): Try all hub combinations
- Updated validate_route() to use hub equivalence helpers

Testing:
- 7 comprehensive test cases covering all edge cases
- 99.56% coverage on tfl_service.py (0 missed lines)
- All 154 existing tests pass (no regressions)

Documentation:
- ADR 07-external-apis.md: Hub interchange validation section
- Created issue #65 for future enhancement (hub IDs as station_tfl_id)

Resolves #52
Related to #47

## Summary by Sourcery

Implement hub-based validation for TfL routes by treating stations with the same hub_naptan_code as interchangeable during connection checks. Refactor the validate_route method into clear, reusable steps, optimize data fetching, enrich logging for hub interchanges, and update documentation with the new behavior.

New Features:
- Support hub-based interchange validation by treating stations sharing the same hub_naptan_code as equivalent
- Introduce helper methods _get_hub_equivalent_stations and _check_any_hub_connection to enable cross-mode station interchanges

Enhancements:
- Refactor validate_route into modular steps with helper functions for segment count, acyclic check, intermediate line IDs, data fetching, and connection validation
- Bulk prefetch stations, lines, and hub-equivalent stations to optimize lookups during route validation
- Enhance structured logging to record hub interchange detection and detailed validation events

Documentation:
- Extend ADR 07 to document hub-based cross-mode interchange validation behavior and architecture

Tests:
- Add comprehensive tests for hub interchange scenarios, edge cases, and newly extracted helper functions, achieving near-complete coverage of route validation logic

### CODE REVIEW

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -1688,35 +1688,40 @@ async def get_line_by_tfl_id(self, tfl_id: str) -> Line:
...
                 False,
                 f"Route cannot have more than {MAX_ROUTE_SEGMENTS} segments. "
                 f"Current route has {len(segments)} segments.",
-                None,
             )
 
-        # Check for duplicate stations (enforce acyclic routes)
+        return True, None
+
+    async def _validate_route_acyclic(self, segments: list[RouteSegmentRequest]) -> tuple[bool, str | None, int | None]:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:1715]: The _validate_route_acyclic function DOES use await self.get_station_by_tfl_id() on line 1734, so it must remain async.
